src/scipp/optimize/__init__.py

```diff
@@ -42,17 +43,29 @@ def _covariance_with_units(pcov, units):
                 u = uj
             elif uj is not None:
                 u = ui * uj
-            pcov[i, j] = _as_scalar(elem, u)
+            pcov[p_names[i]][p_names[j]] = _as_scalar(elem, u)
     return pcov
```

Comment:
I am unsure if the covariance matrix is useful in this format. It is safer than a plain array but don't people usually use it as an actual matrix?

What about returning a DataArray instead?
```.py
pcov = sc.DataArray(data=..., coords={'row': sc.array(dims=['row'], values=p_names), 'col': sc.array(dims=['col'], values=p_names)})
```
Not sure what a good name for the dims would be.

Reply:
> but don't people usually use it as an actual matrix?

This has not been possible before this change either: Since every matrix coefficient typically has a different unit we cannot return an actual matrix (unless the fit function is designed such that each parameter is dimensionless).

> What about returning a DataArray instead?

Same issue with unit... unless you want to return this without units...?

Reply:
Right. Didn't think about that.